### PR TITLE
feat(train): Super-MAS Triton-GPU MAS accelerator (opt-in、 推論影響ゼロ)

### PR DESCRIPTION
**Pull request by ayutaz**
## Summary

学習時の Monotonic Alignment Search (MAS) を Cython から Triton-GPU カーネル (Super-MAS, arXiv:2409.07704, Park et al., 2024) に opt-in で dispatch できるようにしました。 公式 docker stack (Python 3.13 + torch 2.11.0+cu128 + Triton 3.6.0) 上の RTX 4070 Ti SUPER 実測で **MAS 単独 38-187x** 高速化。 ただし MAS は 1 step 全体の **0.42 % しか占有しない** ことが実測で判明したため、 end-to-end の学習短縮効果は **約 0.37 %** とマージナル。 推論経路と ONNX グラフへの影響はゼロ、 既存 Cython は CPU / package 未導入 / 環境変数で disable 時のフォールバックとして温存。

## Affected Components

- [x] Python (src/python/, pyproject.toml)
- [ ] Rust (src/rust/)
- [ ] C# (src/csharp/)
- [ ] C++ (src/cpp/, CMakeLists.txt)
- [ ] Go (src/go/)
- [ ] WASM/npm (src/wasm/)
- [ ] Docker (docker/)
- [ ] CI/CD (.github/workflows/)
- [x] Documentation (docs/, README*)

## Type

- [ ] Bug fix
- [x] New feature
- [ ] Refactoring
- [ ] Documentation
- [ ] CI/CD
- [ ] Dependencies

## Risk Level

- [ ] **patch** — bug fix / internal refactor / docs only
- [x] **minor** — new feature / additive API / no breaking change
- [ ] **major** — breaking change (API removal, schema migration, behavior reversal)

## Contract Impact

- [x] None (Python/runtime-internal change only)

推論経路への影響なし、 ONNX グラフ・cross-runtime contract に変更なし。 学習時 `SynthesizerTrn.forward` 内 1 箇所 (`models.py:964`) のみで dispatch される。

## 変更内容

| 機能名 | 動作 | これがないと起こること |
|---|---|---|
| Super-MAS dispatcher | `super_monotonic_align.maximum_path` が利用可能 + CUDA tensor + `torch.cuda.is_available()` のすべて満たすときのみ Triton kernel に分岐、 それ以外は既存 Cython パスへフォールバック | 学習時 MAS が常に CPU Cython 実装で動作し、 GPU 学習で大規模バッチ時のボトルネックが解消されない |
| upstream の入力破壊への防御 | `_maximum_path_super_mas` 内で `value.clone()` してから kernel へ渡す。 入力 `neg_cent` は不変 | upstream は `value` を in-place 変更するため、 caller の `neg_cent` が壊れて学習 loop が破綻 |
| dtype / device 透過 | float32 化して kernel に渡し、 戻り値を caller の dtype に cast。 device は caller の tensor に従う | Cython パスとの契約 (shape / dtype / device 保持) が崩れ、 既存 callsite と非互換 |
| env による強制 disable | `PIPER_DISABLE_SUPER_MAS=1/true/yes/TRUE/Yes` で import 時に `_super_mas_fn = None` を強制。 再現性デバッグ / bit-identical 比較に利用 | A/B 比較やデバッグ時に Cython へ強制復帰できず原因切り分けができない |
| optional extra (commit SHA pin) | `pyproject.toml` に `super-mas = ["super-monotonic-align @ git+...@<sha>"]` を追加。 unpinned HEAD ではなく **特定 commit SHA** で pin (再現性 + supply-chain hygiene)。 `train` extra には含めず完全 opt-in | デフォルト install で git fetch が走り、 CI / Windows native などネットワーク制約環境で install 失敗。 unpinned だと再現性が崩れ過去 ckpt の bit-identical 再学習不能 |
| dispatcher テスト | `_use_super_mas` predicate / `_maximum_path_super_mas` wrapper / 公開 `maximum_path()` の if 分岐統合点 / env import-time 動作 (subprocess) を網羅。 subprocess test は `os.environ.copy()` で親 env 継承 + `PYTHONPATH` 明示 prepend | dispatcher の wiring 漏れや env が効かない regression を CI が検出できない。 hand-built env では Windows SystemRoot override や piper_train import 失敗のリスク |

## 設計判断

- **3 段安全網で opt-in を確実に**: (1) package 未導入 → Cython、 (2) CPU tensor → Cython (Windows native や CPU-only CI で安全)、 (3) `PIPER_DISABLE_SUPER_MAS=1` → 強制 Cython。 デフォルト挙動は完全に Cython 不変
- **PyPI 非配布のため `train` extra には含めず分離**: upstream は GitHub のみで配布。 git+https を `train` に混ぜると `pip install piper-train[train]` 時にネットワーク fetch が必須となり、 CI / offline 環境で fail する。 `[super-mas]` を独立 extra にすることで「明示的に opt-in した時のみ git+ install」となる
- **commit SHA で pin (HEAD 直追従を回避)**: upstream HEAD は将来移動するため、 unpinned 参照だと過去の学習 run の再現性が壊れる。 SHA pin により kernel revision を学習成果と紐付け、 bump は手動 re-validation 後にのみ行う運用に
- **bit-identical を主張しない**: upstream README は "algorithmic equivalence" のみ表明 (bit-identical は明示保証なし、 加えて `max_neg_val` edge case を修正済)。 docstring / CLAUDE.md でも「algorithmic equivalence」と明記して過剰な主張を避ける
- **`_use_super_mas` を independent predicate に切り出し**: maximum_path() 内に直接書かず関数化することで、 テスト時に predicate のみを mock 可能。 GPU 不在環境でも dispatcher 分岐の wiring を検証できる
- **`value.detach().to(float32).contiguous().clone()`**: upstream は `value` を in-place 変更するため clone 必須。 detach は autograd graph から切る (MAS は学習時 `torch.no_grad()` 配下で呼ばれるが明示する)、 contiguous は Triton kernel が strided memory を扱えない可能性に備える
- **subprocess test は `os.environ.copy()` ベース**: hand-built env では (a) Windows `SystemRoot` を case-insensitive で上書きする可能性、 (b) editable-install されていない piper_train を subprocess から見つけられない、 の 2 リスクがあった。 親 env 継承 + `PYTHONPATH` 明示 prepend で両方解消

## ベンチマーク

計測環境: RTX 4070 Ti SUPER (16GB Ada Lovelace) + WSL2 Ubuntu + Python 3.13.14 + torch 2.11.0+cu128 + Triton 3.6.0 (piper-plus 公式 docker stack と一致)

### MAS 単独 (synthetic, B=20, T=64-400, S=4T)

`triton.testing.do_bench(rep=50, warmup=5)`、 MAS のみの wall-time 比較。

| Shape (B, T, S) | Triton (ms) | Cython (ms) | Speedup |
|---|---:|---:|---:|
| (20, 64, 256) | 0.20 | 7.41 | **37.7x** |
| (20, 128, 512) | 1.00 | 66.61 | **66.5x** |
| (20, 192, 768) | 6.29 | 87.42 | 13.9x (※ outlier) |
| (20, 256, 1024) | 1.32 | 246.98 | **186.8x** |
| (20, 400, 1600) | 3.66 | 417.88 | **114.1x** |

- 頻出する T=64-128 範囲で 38-67x、 大行列 T=256-400 で 114-187x
- T=192 は単発測定の outlier (Triton autotune cache warm-up タイミング依存)

### End-to-end (実 SynthesizerTrn 1 step, B=8, T=128, T_y=512)

forward + backward + optimizer.step 1 step を Triton ON / Cython OFF で各 30 回計測 (warmup 15)。 MAS 呼び出しを patch して MAS 部分のみの wall-time も同時計測。

| 項目 | Triton ON | Cython OFF |
|---|---:|---:|
| MAS median | 1.69 ms | 15.90 ms |
| **MAS 占有率 (step 内)** | **0.027 %** | **0.42 %** |

実測値から計算した end-to-end 短縮: **0.42 % × (1 − 1/9.4) ≈ 0.37 %**

= 1 epoch 8h53m に対して **約 2 分短縮** にとどまる。 これは Glow-TTS 論文の「MAS は学習全体の 2% 未満」推定とも整合的 (実測ではむしろ更に低い 0.42%)。

**つまり: MAS 単独では大幅高速化するが、 学習 step 全体に占める MAS の割合が小さいため、 1 epoch wall-clock の短縮は実用上マージナル**。 大バッチ + 長文 dataset (T=400+) で MAS 占有率がわずかに上がる可能性はあるが、 桁が変わるほどではない。 本 PR の意義は wall-clock 短縮よりも (a) CPU↔GPU 転送オーバーヘッド除去、 (b) 大バッチ時の MAS スパイク解消、 (c) GPU bound にできる範囲を広げる、 にある。

## Test Plan

- [ ] `cd src/python && uv run --no-sync pytest tests/test_monotonic_align.py tests/test_super_mas_dispatch.py --no-cov -v` で 29 件 (既存 8 + 新規 21) すべて pass
- [ ] `uv run --no-sync ruff check src/python/piper_train/vits/monotonic_align/__init__.py src/python/tests/test_super_mas_dispatch.py` で lint clean
- [ ] `uv run --no-sync ruff format --check src/python/piper_train/vits/monotonic_align/__init__.py src/python/tests/test_super_mas_dispatch.py` で format clean
- [ ] 環境変数なしで `python -c "from piper_train.vits import monotonic_align; print(monotonic_align._SUPER_MAS_DISABLED)"` が `False` を返す
- [ ] `PIPER_DISABLE_SUPER_MAS=1` 環境で同コマンドが `True` を返し、 `_super_mas_fn is None` になる
- [ ] (GPU 環境で実施推奨) `pip install "piper-train[super-mas]"` 後に短い学習 step を回し、 既存 Cython 結果と SECS / loss が乖離しないことを確認 (本 PR の merge 後の follow-up でも可)

## Checklist

- [x] Tests pass locally
- [x] No GPL/LGPL dependencies added ([License Policy](CONTRIBUTING.md#license-policy))
- [x] Documentation updated (if applicable)

## Related Issues

なし (`docs/research/architecture-replacement-survey-2026-06-27.md` §4.1.1 で adopt 推奨されていた候補の 1 件目)

### CODE REVIEW

File: src/python/tests/test_super_mas_dispatch.py [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+        import os
+
+        env["PATH"] = os.environ.get("PATH", "")
+        # NOTE: Windows' magic variable is documented as "SystemRoot" (mixed
+        # case). ruff SIM112 prefers the upper-case form but Python's
+        # subprocess on Windows requires the actual name we read from
+        # os.environ, so we explicitly pass the case as-is.
+        sysroot = os.environ.get("SystemRoot")  # noqa: SIM112
+        if sysroot:
+            env["SystemRoot"] = sysroot  # noqa: SIM112
```
[the remark this answers is not in the document]
Comment (ayutaz) [reply src/python/tests/test_super_mas_dispatch.py]: 対応しました (commit 3c6e088c)。 subprocess の env を hand-built dict から `os.environ.copy()` に変更し、 `PYTHONPATH` に `src/python` を明示的に prepend しました。 これにより Windows `SystemRoot` の意図せぬ override を回避しつつ、 piper_train が editable-install されていない環境でも subprocess が import できるようになります。 conftest.py の sys.path injection は pytest 内のみで subprocess には届かないという指摘の通りで、 PYTHONPATH 明示で恒久対応しました。

File: src/python/pyproject.toml [hunk 2]
```diff
@@ -89,6 +89,18 @@ test = [
...
+# CUDA-only. The base Cython MAS remains the default fallback path so this
+# extra is strictly opt-in:
+#
+#     pip install "piper-train[super-mas]"
+#
+# Disable at runtime with PIPER_DISABLE_SUPER_MAS=1 (forces Cython). PyPI
+# distribution does not exist upstream — install is git-only.
+super-mas = [
+    "super-monotonic-align @ git+https://github.com/supertone-inc/super-monotonic-align.git",
+]
```
[the remark this answers is not in the document]
Comment (ayutaz) [reply src/python/pyproject.toml:108]: 対応しました (commit 3c6e088c)。 unpinned HEAD 参照から commit SHA `9bb1cb3a6fbab27bbe6e566827b9548010c5dd51` への pin に切り替えました。 これにより本 extra を使った学習 run の再現性が確保され、 supply-chain reproducibility も担保されます。 SHA bump は今後手動 re-validation (Cython baseline との algorithmic equivalence 確認) を経てから意図的に行う運用にコメントで明記しました。
